apps/web/src/components/AssessmentStepTracker/utils.ts

```diff
@@ -3,17 +3,142 @@ import CheckCircleIcon from "@heroicons/react/20/solid/CheckCircleIcon";
 import PauseCircleIcon from "@heroicons/react/24/solid/PauseCircleIcon";
+import omit from "lodash/omit";
 
 import { IconType } from "@gc-digital-talent/ui";
-
 import {
+  PoolCandidate,
   ArmedForcesStatus,
   AssessmentDecision,
   AssessmentResult,
+  AssessmentResultType,
   Maybe,
-} from "~/api/generated";
+  PoolSkillType,
+} from "@gc-digital-talent/graphql";
+
 import { NO_DECISION, NullableDecision } from "~/utils/assessmentResults";
 
+export type CandidateAssessmentResult = {
+  poolCandidate: PoolCandidate;
+  decision: NullableDecision;
+  results: AssessmentResult[];
+};
+
+/**
+ * Derives the current decision of all assessments
+ * results for a specific set (attributed to a candidate)
+ *
+ * NOTE:
+ *
+ * Essential requirements = essential skills AND education requirements
+ * Optional requirements = asset skills
+ * Assessment Step Status =
+ *  IF any essential requirements unsuccessful THEN "unsuccessful"
+ *  IF any requirements unassessed THEN "to assess"
+ *  IF essential requirements include at least one hold THEN "hold"
+ *  IF all essential requirements successful THEN "success"
+ *
+ * @param results
+ * @returns NullableDecision
+ */
+const deriveResultsDecision = (
+  results: AssessmentResult[],
+): NullableDecision => {
+  // Determine what results are deemed "essential" (essential skills or education requirement)
+  const essentialAssessments = results.filter((result) => {
+    const containsEssentialSkill =
+      result.poolSkill?.type === PoolSkillType.Essential;
+    return (
+      containsEssentialSkill ||
+      result.assessmentResultType === AssessmentResultType.Education
+    );
+  });
+
+  const essentialWasUnsuccessful = !!essentialAssessments.find(
+    (result) => result.assessmentDecision === AssessmentDecision.Unsuccessful,
+  );
+  if (essentialWasUnsuccessful) return AssessmentDecision.Unsuccessful;
+
+  const containsOnHold = !!essentialAssessments.find(
+    (result) => result.assessmentDecision === AssessmentDecision.Hold,
+  );
+  if (containsOnHold) return AssessmentDecision.Hold;
+
+  const containsToAssess = !!results.find(
+    (result) => result.assessmentDecision === null,
+  );
+  if (containsToAssess) return NO_DECISION;
+
+  const essentialWasSuccessful = essentialAssessments.every(
+    (result) => result.assessmentDecision === AssessmentDecision.Successful,
+  );
+  if (essentialWasSuccessful) return AssessmentDecision.Successful;
+
+  return NO_DECISION;
+};
+
+type CandidateAssessmentResultNoDecision = Omit<
+  CandidateAssessmentResult,
+  "decision"
+>;
+
+/**
+ * Given a set of results, group them by candidate
+ * deriving the overall decision from all results
+ * for a specific candidate
+ *
+ * @param results
+ * @returns CandidateAssessmentResult[]
+ */
+export const groupResultsByCandidate = (
+  results: AssessmentResult[],
+): CandidateAssessmentResult[] => {
+  const groupedAssessments = results.reduce<
+    CandidateAssessmentResultNoDecision[]
```

Comment:
Maybe it would be easier (including conceptually) to put this in a Map (keyed by candidate id) instead of using an array and findIndex.